### PR TITLE
[Merged by Bors] - feat(Mathlib.RingTheory.TensorProduct.MvPolynomial) :  tensor product of a (multivariate) polynomial ring

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
Let `Semiring R`, `Algebra R S` and `Module R N`.

* `MvPolynomial.rTensor` gives the linear equivalence
  `MvPolynomial σ S ⊗[R] N ≃ₗ[R] (σ →₀ ℕ) →₀ (S ⊗[R] N)` characterized,
  for `p : MvPolynomial σ S`, `n : N` and `d : σ →₀ ℕ`, by
  `rTensor (p ⊗ₜ[R] n) d = (coeff d p) ⊗ₜ[R] n`
* `MvPolynomial.scalarRTensor` gives the linear equivalence
  `MvPolynomial σ R ⊗[R] N ≃ₗ[R] (σ →₀ ℕ) →₀ N`
  such that `MvPolynomial.scalarRTensor (p ⊗ₜ[R] n) d = coeff d p • n`
  for `p : MvPolynomial σ R`, `n : N` and `d : σ →₀ ℕ`, by

* `MvPolynomial.rTensorAlgHom`, the algebra morphism from the tensor product
  of a polynomial algebra by an algebra to a polynomial algebra
* `MvPolynomial.rTensorAlgEquiv`, `MvPolynomial.scalarRTensorAlgEquiv`,
  the tensor product of a polynomial algebra by an algebra
  is algebraically equivalent to a polynomial algebra

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/MvPolynomial/Basic.lean [hunk 1]
```diff
@@ -645,6 +645,20 @@ def coeffAddMonoidHom (m : σ →₀ ℕ) : MvPolynomial σ R →+ R
   map_add' := coeff_add m
 #align mv_polynomial.coeff_add_monoid_hom MvPolynomial.coeffAddMonoidHom
 
+variable (R)
+/-- `MvPolynomial.coeff m` but promoted to a `LinearMap`. -/
+def lcoeff (m : σ →₀ ℕ) : MvPolynomial σ R →ₗ[R] R where
```
Comment (erdOne) [inline Mathlib/Algebra/MvPolynomial/Basic.lean:650]:
```suggestion
@[simps]
def lcoeff (m : σ →₀ ℕ) : MvPolynomial σ R →ₗ[R] R where
```
Does this generate the lemma below?

File: Mathlib/Data/Finsupp/Basic.lean [hunk 2]
```diff
@@ -214,6 +214,11 @@ def mapRange.addMonoidHom (f : M →+ N) : (α →₀ M) →+ α →₀ N
   map_add' a b := by dsimp only; exact mapRange_add f.map_add _ _; -- Porting note: `dsimp` needed
 #align finsupp.map_range.add_monoid_hom Finsupp.mapRange.addMonoidHom
 
+lemma mapRange.addMonoidHom_apply_single
+    [AddCommMonoid N] [AddCommMonoid P] (e : N →+ P) (a : α) (n : N) :
+    mapRange.addMonoidHom e (single a n) = single a (e n) := by
+    simp only [addMonoidHom_apply, mapRange_single]
```
Comment (erdOne) [inline Mathlib/Data/Finsupp/Basic.lean:220]:
```suggestion
  simp only [addMonoidHom_apply, mapRange_single]
```

Comment (erdOne) [inline Mathlib/Data/Finsupp/Basic.lean:220]: Is this lemma necessary if it is just the composition of two simp lemmas?

File: Mathlib/Data/Finsupp/Defs.lean [hunk 3]
```diff
@@ -455,6 +455,13 @@ theorem unique_single_eq_iff [Unique α] {b' : M} : single a b = single a' b' 
   rw [unique_ext_iff, Unique.eq_default a, Unique.eq_default a', single_eq_same, single_eq_same]
 #align finsupp.unique_single_eq_iff Finsupp.unique_single_eq_iff
 
+lemma apply_single [AddCommMonoid N] [AddCommMonoid P]
+    (e : N →+ P) (a : α) (n : N) (b : α) :
```
Comment (erdOne) [inline Mathlib/Data/Finsupp/Defs.lean:459]: Can you use `MonoidHomClass` here instead?

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+variable {R : Type u} {M : Type v} {N : Type w}
+  [CommSemiring R] [AddCommMonoid M] [Module R M]
+
+variable {σ : Type*} [DecidableEq σ]
+
+variable {S : Type*} [CommSemiring S] [Algebra R S]
+
+section Module
+
+variable   [AddCommMonoid N] [Module R N]
```
Comment (erdOne) [inline Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:51]:
```suggestion
variable [AddCommMonoid N] [Module R N]
```

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    MvPolynomial σ S ⊗[R] N ≃ₗ[S] (σ →₀ ℕ) →₀ (S ⊗[R] N) :=
+  TensorProduct.finsuppLeft' _ _ _ _ _
+
+lemma rTensor_apply_tmul (p : MvPolynomial σ S) (n : N) :
+    rTensor (p ⊗ₜ[R] n) = p.sum (fun i m ↦ Finsupp.single i (m ⊗ₜ[R] n)) :=
+  TensorProduct.finsuppLeft_apply_tmul p n
+
+lemma rTensor_apply_tmul_apply (p : MvPolynomial σ S) (n : N) (d : σ →₀ ℕ) :
+    rTensor (p ⊗ₜ[R] n) d = (coeff d p) ⊗ₜ[R] n :=
+  TensorProduct.finsuppLeft_apply_tmul_apply p n d
```
Comment (erdOne) [inline Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:65]: Should this be a simp lemma?
Also I think there should be `rTensor_apply_monomial_tmul` and `rTensor_apply_X_tmul` as simp lemmas as well.

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+lemma rTensor_apply (t : MvPolynomial σ S ⊗[R] N) (d : σ →₀ ℕ) :
+    rTensor t d = ((lcoeff S d).restrictScalars R).rTensor N t :=
+  TensorProduct.finsuppLeft_apply t d
+
+lemma rTensor_symm_apply_single (d : σ →₀ ℕ) (s : S) (n : N) :
+    rTensor.symm (Finsupp.single d (s ⊗ₜ n)) =
+      (monomial d s) ⊗ₜ[R] n :=
+  TensorProduct.finsuppLeft_symm_apply_single d s n
```
Comment (erdOne) [inline Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:74]: Should this be a simp lemma?

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean (hunk 5 above)
Comment (erdOne) [reply Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:65]: similarly for `scalarRTensor`.

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+/-- The algebra morphism from a tensor product of a polynomial algebra
+  by an algebra to a polynomial algebra -/
+noncomputable def rTensorAlgHom :
+    (MvPolynomial σ S) ⊗[R] N →ₐ[S] MvPolynomial σ (S ⊗[R] N) :=
+  Algebra.TensorProduct.lift
+    (mapAlgHom Algebra.TensorProduct.includeLeft)
+    ((IsScalarTower.toAlgHom R (S ⊗[R] N) _).comp Algebra.TensorProduct.includeRight)
+    (fun p n => by simp [commute_iff_eq, algebraMap_eq, mul_comm])
+
+lemma rTensorAlgHom_coeff_tmul
```
Comment (erdOne) [inline Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:109]:
```suggestion
@[simp]
lemma coeff_rTensorAlgHom_tmul
```
Maybe?

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+  rw [rTensorAlgHom, Algebra.TensorProduct.lift_tmul]
+  rw [AlgHom.coe_comp, IsScalarTower.coe_toAlgHom', Function.comp_apply,
+    Algebra.TensorProduct.includeRight_apply]
+  rw [algebraMap_eq, mul_comm, coeff_C_mul]
+  simp [mapAlgHom, coeff_map]
+
+lemma rTensorAlgHom_toLinearMap :
+    (rTensorAlgHom :
+      MvPolynomial σ S ⊗[R] N →ₐ[S] MvPolynomial σ (S ⊗[R] N)).toLinearMap =
+      (finsuppLeft' _ _ _ _ _).toLinearMap := by
```
Comment (erdOne) [inline Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:121]: Is it possible to use `rTensor` instead of `finsuppLeft'` here?

**Review by erdOne — COMMENTED** [review]
Thanks for splitting the PR up!
I think this is ready for review as well?

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean (hunk 5 above)
Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:65]: I'll do that! thanks for the suggestion.
(Generally speaking, I have little idea of what lemmas should be made simp or not.)

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean (hunk 6 above)
Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:74]: why not

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean (hunk 7 above)
Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:109]: OK.

File: Mathlib/RingTheory/TensorProduct/MvPolynomial.lean (hunk 8 above)
Comment (AntoineChambert-Loir) [reply Mathlib/RingTheory/TensorProduct/MvPolynomial.lean:121]: Yes it was.

File: Mathlib/Algebra/MvPolynomial/Basic.lean (hunk 1 above)
Comment (AntoineChambert-Loir) [reply Mathlib/Algebra/MvPolynomial/Basic.lean:650]: Yes!

File: Mathlib/Data/Finsupp/Basic.lean (hunk 2 above)
Comment (AntoineChambert-Loir) [reply Mathlib/Data/Finsupp/Basic.lean:220]: There's no problem in deleting it if you think this is better. 
And anyway, the simp linter will probably complain one time or another...

File: Mathlib/Data/Finsupp/Defs.lean (hunk 3 above)
Comment (AntoineChambert-Loir) [reply Mathlib/Data/Finsupp/Defs.lean:459]: without any difficulty.

**Review by jcommelin — COMMENTED** [review]
bors d+

**Pull request closed without merging**